api: add unhealthyPodEvictionPolicy for PDBs

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

This PR implements UnhealthyPodEvictionPolicy for PodDisruptionBudgets (originally PodHealthyPolicy) as partially described in https://github.com/kubernetes/enhancements/issues/3017. The additional changes were discussed in this PR.

/kind feature
/kind api-change
/sig apps

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This allows users to specify how PodDisruptionBudget and the eviction API should handle pods that are Running but not Ready.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

- this is a followup to a closed https://github.com/kubernetes/kubernetes/pull/108233
- the implementation part not being considered anymore https://github.com/kubernetes/kubernetes/pull/113376 
- worked on in a collaboration with @mortent  @ravisantoshgudimetla 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
PodDisruptionBudget adds an alpha `spec.unhealthyPodEvictionPolicy` field. When the `PDBUnhealthyPodEvictionPolicy` feature-gate is enabled in `kube-apiserver`, setting this field to `"AlwaysAllow"` allows pods to be evicted if they do not have a ready condition, regardless of whether the PodDisruptionBudget is currently healthy.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3017-pod-healthy-policy-for-pdb
```

atiratree commented on pkg/registry/policy/poddisruptionbudget/strategy.go:
```diff
@@ -167,3 +173,30 @@ func (podDisruptionBudgetStatusStrategy) ValidateUpdate(ctx context.Context, obj
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PDBPodHealthyPolicy) {
+		if podHealthyPolicyInUse(oldPDBSpec) {
+			// prevent updates to the field, so invalid values cannot be set when validation is off
+			pdbSpec.PodHealthyPolicy = oldPDBSpec.PodHealthyPolicy
```
we need this part in case somebody would do `PodRunning` -> `InvalidValue` after we disable the feature, since it also turns off validations for `podHealthyPolicy` field

liggitt replied:
drop this case, validation runs if the field is non-nil regardless of feature gate

atiratree commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -226,10 +228,28 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+				// Make sure the pod status hasn't changed status before we delete.
+				updateDeletionOptions = true
+				return nil
+			}
```
This part needs a proper review/discussion.

In the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md#changes-to-the-eviction-api) we are proposing 

```
The eviction API will be updated to use podHealthyPolicy to determine whether a pod which is Running but not Ready can be evicted regardless of the value of disruptionsAllowed. This will only be a behavioral change when users have specified a podHealthyPolicy, and will not require the actual API to change.
```

Also, @mortent has used a bit different logic in the original PR: https://github.com/mortent/kubernetes/blob/705331a7476a6a596f3122d9b6e3782ade2bfb8b/pkg/registry/core/pod/storage/eviction.go#L226-L241 to implement it like this.

I do not think this is correct, and have changed it as seen above with additional comments as an explanation what is happening in the code.

In case we would go with @mortent solution, we would evict pods without any regard for PDBs which I think is probably not what we want. I am not seeing a benefit of that. I think we want to guard the application pods with a PDB even with other PodHealthyPolicies..

atiratree replied:
I have changed the code and comments to better describe the required behaviour.

atiratree commented on pkg/registry/core/pod/storage/eviction_test.go:
```diff
@@ -31,10 +31,13 @@ import (
+	"k8s.io/kubernetes/pkg/features"
 )
 
 func TestEviction(t *testing.T) {
```
We could also wrap all the tests with the feature flag to test both `disabled`/`enabled` and we could also test policies `""` and `PodReady` here. As mentioned above currently the eviction logic is a copy of the original one, so IMO it should not be that necessary. The price would be more complicated/unsightly tests. 

Any opinions on this?

atiratree replied:
redoing the tests to cover all of the behaviour

leilajal commented:
/triage accepted

atiratree commented on pkg/apis/policy/types.go:
```diff
@@ -42,8 +42,44 @@ type PodDisruptionBudgetSpec struct {
+	// This field is alpha-level. The disruption controller uses this field when
+	// the feature gate PDBPodHealthyPolicy is enabled (disabled by default).
+	// +optional
+	PodHealthyPolicy PodHealthyPolicy
```
I have changed the descriptions as compared to the [KEP](https://github.com/mortent/enhancements/blob/335ce7d92616a68737ba5e63f07316f89741736f/keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md#api). 

I have replaced `legacy` with `default`. I would argue this is not a `legacy` behaviour as it is just a different way of handling a PDB and pods and some people may prefer the original one  to achieve the least disruption to their application. App Running pods have a chance to become Ready, without getting disrupted.  

Also, wouldn't it be better to have eg. a `PodReadyStrict` or `Default` policy. So the behaviour would be on par with the other policies? We could set it on admission by default so we don't have to handle empty strings.

soltysh replied:
Yeah, that's fair, make sure to update the KEP to match this implementation as well.

liggitt replied:
Huh... I wasn't expecting the controller to change at all. I have a hard time seeing a use case for the controller changing to include running-but-not-ready pods in `pdb.Status.CurrentHealthy`.

I thought the option we were adding was strictly around the eviction logic, whether to unconditionally allow eviction of unready pods, regardless of `pdb.Status.CurrentHealthy`. That would avoid changing the controller at all, and limit the entire functional change to:

```diff
 		// If the pod is not ready, it doesn't count towards healthy and we should not decrement
-		if !podutil.IsPodReady(pod) && pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0 {
-			updateDeletionOptions = true
-			return nil
+		if !podutil.IsPodReady(pod) {
+			if utilfeature.DefaultFeatureGate.Enabled(features.UnreadyEvictionPolicy) {
+				if pdb.Spec.UnreadyEvictionPolicy != nil && *pdb.Spec.UnreadyEvictionPolicy == policy.AlwaysAllow {
+					updateDeletionOptions = true
+					return nil
+				}
+			}
+			if pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0 {
+				updateDeletionOptions = true
+				return nil
+			}
 		}
```

Do you have a pointer to the discussion around the decision to shift the design from narrowly modifying the eviction behavior to changing how the controller computes healthy pods?

liggitt replied:
cc folks on the KEP reviews: @wojtek-t, @ravisantoshgudimetla, in addition to @soltysh

liggitt replied:
@soltysh, @atiratree and I synced offline

The `status.currentHealthy` / `status.desiredHealthy` / `status.disruptionsAllowed` fields reflect ready pods today. Redefining those fields to be something other than ready pods seems problematic. Also, none of us could come up with a clear use case for considering running-but-not-ready pods as healthy.

The motivation for this new field and the KEP was to let PDB authors define policies that would not block eviction of unready pods, regardless of the count of healthy pods (to fix things like https://github.com/kubernetes/kubernetes/issues/72320)

We agreed to scope this change down to a field targeting specifically that use case, and to pair it with a narrow change just to the eviction logic for unready pods.

The proposed field is:
```
type UnreadyPodEvictionPolicyType string

// unreadyPodEvictionPolicy determines how pods which are running (status.phase="Running")
// but not yet ready (no status.conditions item with type=Ready,status=True) are handled by the eviction endpoint.
unreadyPodEvictionPolicy *UnreadyPodEvictionPolicyType `json:"...`
```

The allowed values would be:
* `"IfHealthyBudget"` - allow if `pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0` (current behavior)
* `"AlwaysAllow"` - always allow evicting a running-but-unready pod
* `null` - default, same behavior as "AllowWhen

(I don't love the `HealthyBudget` name... that's sort of a placeholder so @atiratree can get started tweaking this PR... if @ravisantoshgudimetla or @wojtek-t or others have a better way to describe the current behavior, we're totally open to it)

atiratree replied:
@liggitt @soltysh 

I have synced with  @ravisantoshgudimetla, and it seems it would be better to change the name of the field from `UnreadyPodEvictionPolicy` -> `UnhealthyPodEvictionPolicy`, to cover future changes to the healthy definition. Users might want to define themselves when their pod becomes healthy. We still will have a binary way of deciding what pod to evict and what not, based on the "healthiness". So I think the current names `IfHealthyBudget` and  `AlwaysAllow` are still fitting. Also, users are used to this naming from the PDB status.

Can I update the name and make the document it in the types in a better way to make it a bit more future proof?

soltysh replied:
If we decide to change healthy definition for PDB it will require much more thorough changes given the current naming in the API.

atiratree replied:
we discussed that it should be probably based on pod readiness gates and we would need to figure out in a new KEP how to make it configurable and backwards compatible with the current solution. But it should be possible with the current changes.

atiratree commented:
/label api-review

soltysh commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -78,6 +80,17 @@ func ValidatePodDisruptionBudgetSpec(spec policy.PodDisruptionBudgetSpec, opts P
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.Selector, labelSelectorValidationOptions, fldPath.Child("selector"))...)
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.PDBPodHealthyPolicy) {
+		supportedPolicies := sets.NewString(
```
It's reasonable to move this var outside the function body, to not to instantiate it every time.

soltysh commented on pkg/api/pod/util.go:
```diff
@@ -233,6 +233,11 @@ func visitContainerConfigmapNames(container *api.Container, visitor Visitor) boo
 	return true
 }
 
+// IsPodRunning returns true if the phase of the pod is PodRunning; false otherwise.
```
I'd argue that this check is so simple we don't need to create a helper here, it's different when looking at conditions. Ie, I'd drop this helper.

soltysh commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -78,6 +80,17 @@ func ValidatePodDisruptionBudgetSpec(spec policy.PodDisruptionBudgetSpec, opts P
 
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.Selector, labelSelectorValidationOptions, fldPath.Child("selector"))...)
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.PDBPodHealthyPolicy) {
```
You want to validate the values always, otherwise you'll end up with a resource with invalid values which then won't be able to run when the feature is on.

soltysh commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -226,10 +226,33 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+				updateDeletionOptions = true
+				return nil
+			}
+		case pdb.Spec.PodHealthyPolicy == policyv1.PodReady:
```
`PodReady` is identical with empty `""` policy, so I'd expect the condition be identical or squash this condition with above, right?

atiratree replied:
That is the misconception that can arise... Please see https://github.com/kubernetes/kubernetes/pull/113375#discussion_r1006321318 and the comments here.

The difference is in `&& pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0` as this guards the application until the `DesiredHealthy` number is reached. That is Running pods have a chance to become ready.

liggitt commented on pkg/registry/policy/poddisruptionbudget/strategy.go:
```diff
@@ -182,3 +188,26 @@ func hasInvalidLabelValueInLabelSelector(pdb *policy.PodDisruptionBudget) bool {
+			// prevent updates to the field, so invalid values cannot be set when validation is off
+			pdbSpec.PodHealthyPolicy = oldPDBSpec.PodHealthyPolicy
+		} else {
+			pdbSpec.PodHealthyPolicy = ""
```
set to nil here

liggitt commented:
I think https://github.com/kubernetes/kubernetes/pull/113375#discussion_r1016962746 is the comment we should address first, since it shifts the rest of the PR... the last time I interacted with the goal for this feature, it was in the context of https://github.com/kubernetes/kubernetes/pull/105296/files#diff-6105105bc4d40e3d45b9cc38165516be0205f0329929421fd7df65df7abaeb2cL223, and it's not clear to me that either of the options this design adds fixes the problem that was trying to address

leonardpahlke commented:
exception request approved: [email](https://groups.google.com/a/kubernetes.io/g/release-team/c/UYrZPLI4mdU/m/e09U65wmAQAJ)

Your updated deadline to make any changes to your PR is [18:00 PST Friday 11th November 2022](https://everytimezone.com/s/042f76f9).

/milestone v1.26

atiratree committed:
integration: add eviction tests for UnhealthyPodEvictionPolicy

atiratree committed:
api: add UnhealthyPodEvictionPolicy for PDBs

atiratree commented:
@liggitt @soltysh I have updated the PR and moved the remaining bits from https://github.com/kubernetes/kubernetes/pull/113376 to here (integration tests), so this is the only PR now. Can you please review and give feedback to this https://github.com/kubernetes/kubernetes/pull/113375#discussion_r1019308856 ?

atiratree committed:
generated

soltysh commented:
/priority backlog

soltysh reviewed (approved):
/lgtm
/approve
from sig-apps pov

atiratree commented:
`should not create extra sandbox if all containers are done` seems irrelevant and probably a flake 
/test pull-kubernetes-e2e-kind-ipv6

liggitt commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -226,10 +226,26 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
-			return nil
+		// IsPodReady is the current implementation of IsHealthy
+		// If the pod is healthy, it should be guarded by the PDB.
+		// If the pod is not healthy, it doesn't count towards currentHealthy and desiredHealthy and we should not decrement disruptionsAllowed.
```
move the `it doesn't count towards currentHealthy and desiredHealthy and we should not decrement disruptionsAllowed` comment to the two places inside this block where we return nil... when we don't return and continue on, we *do* decrement `disruptionsAllowed`

atiratree replied:
+1, changed

liggitt reviewed (commented):
API mechanics lgtm, some doc/comment updates needed

atiratree commented on pkg/apis/policy/types.go:
```diff
@@ -42,8 +42,37 @@ type PodDisruptionBudgetSpec struct {
+	// (no status.conditions item with type=Ready,status=True) should be
+	// considered for eviction. If no policy is specified, the default behavior
+	// will be used, which corresponds to the IfHealthyBudget policy.
+	//
```
done, but not sure if the format is correct

also, I think It would be preferable for clients to fail when encountering unrecognized policy. Does this apply to eviction API as well? Should I change the code to fail if policy is unrecognized? 

@liggitt thoughts?

liggitt replied:
I guess clients that just read/display this field just need to be prepared to display additional values.

I agree clients making eviction decisions based on this should default to disallowing eviction of an unready pod if they encounter an unknown policy value. (the only current client making these decisions is the apiserver itself, which will never encounter an unknown value for this field)

how about:
```
	// Additional policies may be added in the future.
  // Clients making eviction decisions should disallow eviction of unready pods if they encounter an unrecognized policy in this field.
```

atiratree replied:
Sounds good - will update the wording. And keep the eviction implementation as is, since we are in control of changing it when we add new policies.

atiratree commented on pkg/apis/policy/types.go:
```diff
@@ -42,8 +42,38 @@ type PodDisruptionBudgetSpec struct {
+	// should be considered for eviction. Current implementation considers healthy pods,
+	// as pods that have status.conditions item with type=Ready,status=True.
+	// If no policy is specified, the default behavior will be used,
+	// which corresponds to the IfHealthyBudget policy.
```
noted, will add to the KEP update

atiratree commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -226,10 +226,26 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+				if pdb.Spec.UnhealthyPodEvictionPolicy != nil && *pdb.Spec.UnhealthyPodEvictionPolicy == policyv1.AlwaysAllow {
+					// Note: deleting unhealthy (unready) pods can cause inability of the disrupted application to come back up,
+					// since its perspective running pods can be evicted without consulting the PDB,
+					// this applies to the new pods and also to the pods that have been healthy but became healthy after.
```
removed from here, and changed the wording a bit in the API

atiratree commented:
flaky `run_deprecated_api_tests`
/test pull-kubernetes-integration

liggitt commented:
/lgtm
/approve

atiratree commented:
unrelated `Networking Granular Checks` flaking
/test pull-kubernetes-e2e-kind-ipv6

atiratree commented:
same ^ 
/test pull-kubernetes-e2e-kind-ipv6
